File: src/main/java/de/visustruct/control/CanvasStyle.java
package de.visustruct.control;

import java.awt.Color;

import javax.swing.LookAndFeel;
import javax.swing.UIManager;

import com.formdev.flatlaf.FlatDarkLaf;
import com.formdev.flatlaf.themes.FlatMacDarkLaf;

/**
 * Farben der Struktogramm-Zeichenfläche (nicht die Swing-Fenster).
 * Bei <b>Modern · dark</b> (FlatDark) wird eine hellere graue Fläche genutzt, passend zum dunklen UI.
 * <p>
 * Helles Schema: an {@code VisuStruct-SwiftUI} / {@code DiagramCanvasView} angelehnt (weiße Zellen,
 * schwarze Kanten 2 px, blauer Auswahlrahmen).
 */
public final class CanvasStyle {

	/** Kanten von Blöcken und Hilfslinien — entspricht Swift {@code DiagramCanvasInk.diagramLineWidth}. */
	public static final float DIAGRAM_LINE_WIDTH = 2f;
	/** Auswahlrahmen um die aktive Kachel — entspricht Swift {@code DiagramCanvasInk.selectionLineWidth}. */
	public static final float SELECTION_LINE_WIDTH = 3.5f;

	private static Color background;
	private static Color diagramFrame;
	private static Color titleText;
	private static Color elementBorder;
	private static Color elementText;
	private static Color elementSelectedFill;
	private static Color selectionStroke;
	private static Color dropPreview;
	private static Color dragFrame;
	private static Color elementFill;
	private static Color simulationStepHighlightFill;

	private CanvasStyle() {
	}

	static {
		applyLightPalette();
	}

	private static void applyLightPalette() {
		// Canvas ~ Swift Color(white: 0.98); Rahmen/Kanten schwarz wie im Swift-Diagramm-Canvas.
		background = new Color(0xFAFAFA);
		diagramFrame = Color.BLACK;
		titleText = new Color(0x111827);
		elementBorder = Color.BLACK;
		elementText = new Color(0x111827);
		// Swift: Auswahl = weiße Füllung + blauer Rand; Füllung kommt aus {@code elementFill}, Rahmen aus {@code selectionStroke}.
		elementSelectedFill = new Color(0xDBEAFE);
		selectionStroke = new Color(0f, 0.48f, 1f, 0.9f);
		// Drop-Vorschau: kräftiger + halbtransparent, damit sie über Diagrammflächen nicht "klebt".
		dropPreview = new Color(0x803B82F6, true);
		dragFrame = new Color(0x3B82F6);
		elementFill = Color.WHITE;
		simulationStepHighlightFill = new Color(0xBFDBFE);
	}

	/** Zeichenfläche spürbar dunkler; Blöcke bleiben hell genug für Lesbarkeit. */
	private static void applyDarkPalette() {
		background = new Color(0x1E1E22);
		diagramFrame = new Color(0xB4B4BA);
		titleText = new Color(0xF4F4F5);
		elementBorder = new Color(0xB4B4BA);
		elementText = new Color(0x18181B);
		elementSelectedFill = new Color(0xA7C7FF);
		selectionStroke = new Color(0.35f, 0.63f, 1f, 0.92f);
		// Drop-Vorschau: stärker + halbtransparent.
		dropPreview = new Color(0x805B9FFF, true);
		dragFrame = new Color(0x5B9FFF);
		elementFill = new Color(0xECECF0);
		simulationStepHighlightFill = new Color(0x1E3A5F);
	}

	private static boolean isFlatDarkTheme() {
		LookAndFeel laf = UIManager.getLookAndFeel();
		return laf instanceof FlatDarkLaf || laf instanceof FlatMacDarkLaf;
	}

	/**
	 * Nach Theme-Wechsel ({@link javax.swing.UIManager#setLookAndFeel}) aufrufen, z. B. aus {@link de.visustruct.view.UiTheme#applyAfterTheme()}.
	 */
	public static void syncToTheme() {
		if (isFlatDarkTheme()) {
			applyDarkPalette();
		} else {
			applyLightPalette();
		}
	}

	public static Color getBackground() {
		return background;
	}

	public static Color getDiagramFrame() {
		return diagramFrame;
	}

	public static Color getTitleText() {
		return titleText;
	}

	public static Color getElementBorder() {
		return elementBorder;
	}

	public static Color getElementText() {
		return elementText;
	}

	public static Color getElementSelectedFill() {
		return elementSelectedFill;
	}

	/** Randfarbe für die aktuell ausgewählte Kachel (Swift: {@code selectionStrokeColor}). */
	public static Color getSelectionStroke() {
		return selectionStroke;
	}

	public static Color getDropPreview() {
		return dropPreview;
	}

	public static Color getDragFrame() {
		return dragFrame;
	}

	/** Füllfarbe der Blöcke (nicht markiert). */
	public static Color getElementFill() {
		return elementFill;
	}

	/** Hintergrund für den aktuellen Simulations-Schritt im Diagramm. */
	public static Color getSimulationStepHighlightFill() {
		return simulationStepHighlightFill;
	}
}

File: src/main/kotlin/de/visustruct/control/CanvasStyle.kt
package de.visustruct.control

import com.formdev.flatlaf.FlatDarkLaf
import com.formdev.flatlaf.themes.FlatMacDarkLaf
import java.awt.Color
import javax.swing.LookAndFeel
import javax.swing.UIManager

/**
 * Farben der Struktogramm-Zeichenfläche (nicht die Swing-Fenster).
 */
object CanvasStyle {

    const val DIAGRAM_LINE_WIDTH: Float = 2f
    const val SELECTION_LINE_WIDTH: Float = 3.5f

    private var background: Color = Color(0xFAFAFA)
    private var diagramFrame: Color = Color.BLACK
    private var titleText: Color = Color(0x111827)
    private var elementBorder: Color = Color.BLACK
    private var elementText: Color = Color(0x111827)
    private var elementSelectedFill: Color = Color(0xDBEAFE)
    private var selectionStroke: Color = Color(0f, 0.48f, 1f, 0.9f)
    private var dropPreview: Color = Color(0x803B82F6.toInt(), true)
    private var dragFrame: Color = Color(0x3B82F6)
    private var elementFill: Color = Color.WHITE
    private var simulationStepHighlightFill: Color = Color(0xBFDBFE)

    init {
        applyLightPalette()
    }

    private fun applyLightPalette() {
        background = Color(0xFAFAFA)
        diagramFrame = Color.BLACK
        titleText = Color(0x111827)
        elementBorder = Color.BLACK
        elementText = Color(0x111827)
        elementSelectedFill = Color(0xDBEAFE)
        selectionStroke = Color(0f, 0.48f, 1f, 0.9f)
        dropPreview = Color(0x803B82F6.toInt(), true)
        dragFrame = Color(0x3B82F6)
        elementFill = Color.WHITE
        simulationStepHighlightFill = Color(0xBFDBFE)
    }

    private fun applyDarkPalette() {
        background = Color(0x1E1E22)
        diagramFrame = Color(0xB4B4BA)
        titleText = Color(0xF4F4F5)
        elementBorder = Color(0xB4B4BA)
        elementText = Color(0x18181B)
        elementSelectedFill = Color(0xA7C7FF)
        selectionStroke = Color(0.35f, 0.63f, 1f, 0.92f)
        dropPreview = Color(0x805B9FFF.toInt(), true)
        dragFrame = Color(0x5B9FFF)
        elementFill = Color(0xECECF0)
        simulationStepHighlightFill = Color(0x1E3A5F)
    }

    private fun isFlatDarkTheme(): Boolean {
        val laf: LookAndFeel = UIManager.getLookAndFeel()
        return laf is FlatDarkLaf || laf is FlatMacDarkLaf
    }

    @JvmStatic
    fun syncToTheme() {
        if (isFlatDarkTheme()) {
            applyDarkPalette()
        } else {
            applyLightPalette()
        }
    }

    @JvmStatic
    fun getBackground(): Color = background

    @JvmStatic
    fun getDiagramFrame(): Color = diagramFrame

    @JvmStatic
    fun getTitleText(): Color = titleText

    @JvmStatic
    fun getElementBorder(): Color = elementBorder

    @JvmStatic
    fun getElementText(): Color = elementText

    @JvmStatic
    fun getElementSelectedFill(): Color = elementSelectedFill

    @JvmStatic
    fun getSelectionStroke(): Color = selectionStroke

    @JvmStatic
    fun getDropPreview(): Color = dropPreview

    @JvmStatic
    fun getDragFrame(): Color = dragFrame

    @JvmStatic
    fun getElementFill(): Color = elementFill

    @JvmStatic
    fun getSimulationStepHighlightFill(): Color = simulationStepHighlightFill
}
